group-coordinator/src/test/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroupTest.java

```diff
@@ -0,0 +1,500 @@
+        assertEquals(-1, consumerGroup.currentPartitionEpoch(fooTopicId, 9));
+    }
+
+    @Test
+    public void testGroupState() {
+        ConsumerGroup consumerGroup = createConsumerGroup("foo");
+        assertEquals(ConsumerGroup.ConsumerGroupState.EMPTY, consumerGroup.state());
+
+        ConsumerGroupMember member1 = new ConsumerGroupMember.Builder("member1")
+            .setMemberEpoch(1)
+            .setPreviousMemberEpoch(0)
+            .setNextMemberEpoch(1)
+            .build();
+
+        consumerGroup.updateMember(member1);
+        consumerGroup.setGroupEpoch(1);
+
+        assertEquals(ConsumerGroup.ConsumerGroupState.ASSIGNING, consumerGroup.state());
+
+        ConsumerGroupMember member2 = new ConsumerGroupMember.Builder("member2")
+            .setMemberEpoch(1)
+            .setPreviousMemberEpoch(0)
+            .setNextMemberEpoch(1)
+            .build();
+
+        consumerGroup.updateMember(member2);
+        consumerGroup.setGroupEpoch(2);
+
+        assertEquals(ConsumerGroup.ConsumerGroupState.ASSIGNING, consumerGroup.state());
+
+        consumerGroup.setAssignmentEpoch(2);
+
+        assertEquals(ConsumerGroup.ConsumerGroupState.RECONCILING, consumerGroup.state());
+
+        member1 = new ConsumerGroupMember.Builder(member1)
+            .setMemberEpoch(2)
+            .setPreviousMemberEpoch(1)
+            .setNextMemberEpoch(2)
+            .build();
+        consumerGroup.updateMember(member1);
+
+        assertEquals(ConsumerGroup.ConsumerGroupState.RECONCILING, consumerGroup.state());
+
+        member2 = new ConsumerGroupMember.Builder(member2)
+            .setMemberEpoch(2)
+            .setPreviousMemberEpoch(1)
+            .setNextMemberEpoch(2)
+            .build();
+        consumerGroup.updateMember(member2);
+
+        assertEquals(ConsumerGroup.ConsumerGroupState.STABLE, consumerGroup.state());
+
+        consumerGroup.removeMember("member1");
+        consumerGroup.removeMember("member2");
+
+        assertEquals(ConsumerGroup.ConsumerGroupState.EMPTY, consumerGroup.state());
+    }
+
+    @Test
+    public void testPreferredServerAssignor() {
+        ConsumerGroup consumerGroup = createConsumerGroup("foo");
+
+        ConsumerGroupMember member1 = new ConsumerGroupMember.Builder("member1")
+            .setServerAssignorName("range")
+            .build();
+        ConsumerGroupMember member2 = new ConsumerGroupMember.Builder("member2")
+            .setServerAssignorName("range")
+            .build();
+        ConsumerGroupMember member3 = new ConsumerGroupMember.Builder("member3")
+            .setServerAssignorName("uniform")
+            .build();
+
+        assertEquals(
+            Optional.empty(),
+            consumerGroup.preferredServerAssignor(null, null)
+        );
+
+        assertEquals(
+            Optional.of("range"),
+            consumerGroup.preferredServerAssignor(null, member1)
+        );
+
+        consumerGroup.updateMember(member1);
```

Comment:
So this is basically a no-op since we subtract 1 for the previous member and add one for the new member.

Reply:
Actually I'm confused. We should have null as the current member, so shouldn't the count be 2 for range?